stdlib/src/builtin/sort.mojo

```diff
@@ -15,400 +15,169 @@
-    The ordering of the first k elements is undefined.
+fn _quick_sort[D: DType](inout list: List[Scalar[D]], low: Int, high: Int):
+    """Sort section of the list, between low and high, with quick sort algorithm inplace.
 
     Parameters:
-        type: DType of the underlying data.
-        cmp_fn: Comparison functor of type, type) capturing -> Bool type.
-
-    Args:
-        buff: Input buffer.
-        k: Index of the partition element.
-        size: The length of the buffer.
-    """
-    var stack = List[Int](capacity=_estimate_initial_height(size))
-    stack.append(0)
-    stack.append(size)
-    while len(stack) > 0:
-        var end = stack.pop()
-        var start = stack.pop()
-        var pivot = _partition[type, cmp_fn](buff, start, end)
-        if pivot == k:
-            break
-        elif k < pivot:
-            stack.append(start)
-            stack.append(pivot)
-        else:
-            stack.append(pivot + 1)
-            stack.append(end)
-
-
-# ===----------------------------------------------------------------------===#
-# sort
-# ===----------------------------------------------------------------------===#
-
-
-fn sort(inout buff: Pointer[Int], len: Int):
-    """Sort the vector inplace.
-    The function doesn't return anything, the vector is updated inplace.
+        D: The dtype of the scalar.
 
     Args:
-        buff: Input buffer.
-        len: The length of the buffer.
+        list: The list of the scalars which will be sorted inpace.
+        low: Int value identifying the lowes index of the list section to be sorted.
+        high: Int value identifying the highest index of the list section to be sorted.
     """
 
+    @always_inline
     @parameter
-    fn _less_than_equal[type: AnyRegType](lhs: type, rhs: type) -> Bool:
-        return rebind[Int](lhs) <= rebind[Int](rhs)
-
-    _quicksort[Int, _less_than_equal](buff, len)
+    fn _partition(low: Int, high: Int) -> Int:
+        var pivot = list[high]
+        var i = low - 1
+        for j in range(low, high):
+            if list[j] <= pivot:
+                i += 1
+                list[j], list[i] = list[i], list[j]
+        list[i + 1], list[high] = list[high], list[i + 1]
+        return i + 1
+
+    if low < high:
+        var pi = _partition(low, high)
+        _quick_sort(list, low, pi - 1)
+        _quick_sort(list, pi + 1, high)
 
 
-fn sort[type: DType](inout buff: Pointer[Scalar[type]], len: Int):
-    """Sort the vector inplace.
-    The function doesn't return anything, the vector is updated inplace.
+@always_inline
+fn quick_sort[D: DType](inout list: List[Scalar[D]]):
+    """Sort list of scalars in place with quick sort algorithm.
 
     Parameters:
-        type: DType of the underlying data.
+        D: The dtype of the scalar.
 
     Args:
-        buff: Input buffer.
-        len: The length of the buffer.
+        list: The list of the scalars which will be sorted inpace.
     """
+    _quick_sort(list, 0, len(list) - 1)
```

Comment:
**Question** Do we have tests for `quick_sort` on zero-length lists?  Anytime I see `len(list) - 1`, I worry about the end bounds, but I suspect `_quick_sort` does the right thing still.